python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll

```diff
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
   )
 }
 
+/**
+ * The unpacking assignment takes the general form
+ * ```python
+ *   sequence = iterable
+ * ```
+ * where `sequence` is either a tuple or a list and it can contain wildcards.
+ * The iterable can be any iterable, which means that content will need to change type
+ * if it should be transferred from the LHS to the RHS.
+ *
+ * We may for instance have
+ * ```python
+ *    (a, b) = ["a", "tainted string"]  # RHS has content `ListElement`
+ * ```
+ * Due to the abstraction for list content, we do not know whether `"tainted string"`
+ * ends up in `a` or in `b`, so we want to overapproximate and see it in both.
+ *
+ * Using wildcards we may have
+ * ```python
+ *   (a, *b) = ("a", "b", "tainted string")  # RHS has content `TupleElement(2)`
+ * ```
+ * Since the starred variables are always assigned type list, `*b` will be
+ * `["b", "tainted string]`, and we will agsin overapproximate and assign it
+ * content corresponding to anything found in the RHS.
+ *
+ * For a precise transfer
+ * ```python
+ *    (a, b) = ("a", "tainted string")  # RHS has content `TupleElement(1)`
+ * ```
+ * we wish to keep the precision, so only `b` receives the tuple content at index 1.
+ *
+ * Finally, `sequence` is actually a pattern and can have a more complicated structure,
+ * such as
+ * ```python
+ *   (a, [b, *c]) = ("a", ("tainted string", "c"))  # RHS has content `TupleElement(1); TupleElement(0)`
+ * ```
+ * where `a` should not receive content, but `b` and `c` should. `c` will be `["c"]` so
+ * should have the content converted and transferred, while `b` should read it.
```

Comment:
In this case `b` and `c` are both inside a list so if there is any content in the tuple matched by `[b, *c]` then `[b, *c]` should have content `ListElement [rest of acces path]` and `b` should have `[rest of acces path]` and `c` should have `ListElement [rest of acces path]`.

Reply:
Hmm, my understanding of iterable unpacking is that it doesn't matter if you use normal parentheses or square brackets to match the pattern (but that having both can enable writing clean code). So

```py
a, (b, c) = iterable
(a, (b, c)) = iterable
[a, [b, c]] = iterable
a, [b, c] = iterable
```

are all semantically equivalent -- or have I misunderstood something?

Reply:
I would say that, apart from the first two, these are all different. But the difference is only in the types of intermediate results, so you only see it if that difference matters. It might for instance matter for runtime. It matters for us, because we distinguish list-content and tuple-content. In terms of the values of `a`, `b`, and `c`, the interpreter does not see a difference because the third element of a list is just as precisely known as the third element of a tuple and so it can move back and forth between the two without noticing. Our analysis does see a difference, which is why it does not agree with the interpreter on the following program:
```python
  l = [SOURCE, NONSOURCE]
  t = (l[0], l[1])
```
The interpreter sees `t` having `SOURCE` only in the first component, our analysis says it can be found in both.

Reply:
So, just to be perfectly clear, do you agree that these are all semantically equivalent? (in terms of Python semantics)

Reply:
What is Python semantics? They may incur different runtimes or memory footprints, and if you manage to change the subscripting of either list or tuple, they could also give different values.

Reply:
What is anything, really? :)

My point here is that we could make our analysis a bit simpler, by ignoring whether square brackets or normal parenthesis was used on the LHS of the assignment. I'm guessing that with our current Python libraries, we represent these as `TupleNode` or `ListNode`, but I don't think we need to make the distinction here.

After spending some time looking for evidence on this, I finally found https://docs.python.org/3/reference/simple_stmts.html#assignment-statements:

> Assignment of an object to a target list, optionally enclosed in parentheses or square brackets, is recursively defined as follows.

So if a target is actually a target list (to use the wording from that Python reference), I think we can just consider that target list to be a tuple. This will allow us to simplify the data-flow modeling a bit, and might lead to a bit more precision in some cases (for example `(a, [b, *c]) = ("a", ("tainted string", "c"))`).

### non-goals

I'm not suggesting we should change the behavior of our data-flow modeling in general. So in the example below, when tracking data-flow from `SOURCE`, I'm perfectly fine with modeling content of `l` as _some element_, meaning that we think the content might flow to `b`. If we ever want to change that, this is not the PR to do it in at least :smile: 

```py
l = [SOURCE, NON_SOURCE]
t = (SOURCE, NON_SOURCE)

a, b = l
x, y = t
```

Reply:
> What is anything, really? :)

I meant that one could define many semantics for Python, depending on ones interest. One such might or might not include expected runtime or memory footprint.

> After spending some time looking for evidence on this, I finally found https://docs.python.org/3/reference/simple_stmts.html#assignment-statements:
> 
> > Assignment of an object to a target list, optionally enclosed in parentheses or square brackets, is recursively defined as follows.
> 
> So if a target is actually a target list (to use the wording from that Python reference), I think we can just consider that target list to be a tuple. This will allow us to simplify the data-flow modeling a bit, and might lead to a bit more precision in some cases (for example `(a, [b, *c]) = ("a", ("tainted string", "c"))`).

Indeed, it looks like you are right! The different notations are just there for convenience and always denote a "target list", which can indeed model as a tuple.

Reading the reference, it also occurred to me that we may not handle the assignment (and the possible call to a special method) hidden in
```python
  a, b[0] = 1, 2
```

Reply:
```py
 a, b[0] = 1, 2
```

\*goes to check in ipython\* ... good catch! Would be great with a test-case at least, but not sure if you want to handle that (more general) case in this PR?

Reply:
I think adding a test case is probably a good idea, but I don't think fixing this needs to be done in this PR.
Also, it seems like this is two different issues:

1. Correct handling of content when assigning to a subscripted expression, and
2. correctly handling the case where `__setitem__` exists on `b`.

Of these, I would say 1. is probably the higher priority (though there may exist fancy libraries that use `__setitem__` to do clever stuff, so we may end up needing both).